Parse document correctly when AMP emoji is used

schlessera opened the pull request:
Fixes #75

westonruter commented on src/Dom/Document.php:
```diff
@@ -1099,6 +1110,38 @@ private function convertAmpBindAttributes($html)
         return (null !== $converted) ? $converted : $html;
     }
 
+    /**
+     * Convert AMP bind-attributes back to their original syntax.
+     *
+     * This is not guaranteed to produce the exact same result as the initial markup, as it is more of a best guess.
+     * It can end up replacing the wrong attributes if the initial markup had inconsistent styling, mixing both syntaxes
+     * for the same attribute. In either case, it will always produce working markup, so this is not that big of a deal.
+     *
+     * @see convertAmpBindAttributes() Reciprocal function.
+     * @link https://www.ampproject.org/docs/reference/components/amp-bind
+     *
+     * @param string $html HTML with amp-bind attributes converted.
+     * @return string HTML with amp-bind attributes restored.
+     */
+    public function restoreAmpBindAttributes($html)
+    {
+        if (empty($this->convertedAmpBindAttributes)) {
+            return $html;
+        }
+
+        $pattern     = sprintf(
+            '#%s(%s)#i',
+            self::AMP_BIND_DATA_ATTR_PREFIX,
+            implode('|', array_unique($this->convertedAmpBindAttributes))
+        );
+        $replacement = '[$1]';
+        $limit       = count($this->convertedAmpBindAttributes);
+
+        $restored = preg_replace($pattern, $replacement, $html, $limit);
+
+        return (null !== $restored) ? $restored : $html;
+    }
```
I still don't believe that _restoring_ is the right approach here since attributes can be added/removed after the DOM has been parsed and before it has been serialized, resulting in a incorrect restoration as the counts will change: https://github.com/ampproject/amp-toolbox-php/commit/bb607f85ae3e277d67627d03b9e0ea80f6ab2476#commitcomment-47328593

As mentioned in https://github.com/ampproject/amp-toolbox-php/commit/bb607f85ae3e277d67627d03b9e0ea80f6ab2476#commitcomment-47328668:

> Therefore, I think the only reasonable path is to introduce a `useBracketedAmpBindSyntax` flag which controls whether _all_ attributes are serialized using the bracketed syntax or the `data-amp-bind-` syntax. By default it would be `false`, but in unit tests you could make it `true`.

schlessera replied:
So, what I did now is to provide a way to pass options around the document (with BC fallback that mimic the built-in `DOMDocument` do its syntax still works) and I added an option (`Document::OPTION_AMP_BIND_SYNTAX`) to configure how the conversion should be handled for `amp-bind` attributes. There are three available options:
- `Document::AMP_BIND_SYNTAX_AUTO` => try a best guess to keep the individual attributes in the syntax they were in originally.
- `Document::AMP_BIND_SYNTAX_DATA_ATTRIBUTE` => convert all `amp-bind` syntax into the `data-amp-bind-*` syntax.
- `Document::AMP_BIND_SYNTAX_SQUARE_BRACKETS` => convert all `amp-bind` attributes into the `[*]` syntax.

We can then set the option to `Document::AMP_BIND_SYNTAX_DATA_ATTRIBUTE` in the WordPress plugin to keep the current behavior for WP unchanged.

The default is `Document::AMP_BIND_SYNTAX_AUTO`, as I think this is the expected behavior, as third-party users of the library would not even be aware the `DOMDocument` code is broken and requires such a behind-the-scenes conversion in the first place.

westonruter commented on src/Dom/Document.php:
```diff
@@ -327,15 +379,23 @@ public function __construct($version = '', $encoding = null)
+            $options = [self::OPTION_ENCODING => $options];
+        }
+
+        $encoding = isset($options[self::OPTION_ENCODING]) ? $options[self::OPTION_ENCODING] : null;
+
         $dom = new self('', $encoding);
```
Alternatively:

```suggestion
        $options = array_merge( self::KNOWN_OPTIONS, $options );

        $dom = new self('', $options[self::OPTION_ENCODING]);
```

westonruter replied:
But I see this merge is already being done below.

schlessera replied:
Yes, I wanted to avoid multiple merges.

westonruter commented on src/Dom/Document.php:
```diff
@@ -347,15 +407,23 @@ public static function fromHtml($html, $encoding = null)
+            $options = [self::OPTION_ENCODING => $options];
+        }
+
+        $encoding = isset($options[self::OPTION_ENCODING]) ? $options[self::OPTION_ENCODING] : null;
+
         $dom = new self('', $encoding);
```
```suggestion
        $options = array_merge( self::KNOWN_OPTIONS, $options );

        $dom = new self('', $options[self::OPTION_ENCODING]);
```

westonruter replied:
Ditto above, but nevermind per below.

westonruter commented on src/Dom/Document.php:
```diff
@@ -1099,6 +1183,54 @@ private function convertAmpBindAttributes($html)
+            return $html;
+        }
+
+        // Depending on options, restore previously converted or all amp-bind attributes to square bracket syntax.
+        $pattern = $this->options[self::OPTION_AMP_BIND_SYNTAX] === self::AMP_BIND_SYNTAX_AUTO
+            ? sprintf(
+                '#%s(%s)#i',
+                self::AMP_BIND_DATA_ATTR_PREFIX,
+                implode('|', array_unique($this->convertedAmpBindAttributes))
+            )
+            : sprintf(
+                '#%s([a-z0-9-_]+)#i', // TODO: Check what exact regex makes sense here.
+                self::AMP_BIND_DATA_ATTR_PREFIX
+            );
```
Don't the patterns here run the risk of unintentional conversion? For example, if I write a post about amp-bind and I say:

> You have two options: use the `[text]` attribute or the `data-amp-bind-text` attribute.

At present this will result in the unexpected:

> You have two options: use the `[text]` attribute or the `[text]` attribute.

In order to safely convert back to the bracketed syntax, we'd need to use the same logic in `convertAmpBindAttributes` which uses the `AMP_BIND_START_TAGS_PATTERN` regex to first match all tags which contain an attribute in question, and then in the replacement callback do the actual replacements.

schlessera replied:
Yes, indeed, this needs to become way more robust.

schlessera replied:
I think this should now work as expected (see added tests with latest commits).

The regex I ended up with looks overly complex, so happy to hear feedback if I'm missing something obvious.

westonruter commented on src/Dom/Document.php:
```diff
@@ -1082,7 +1201,93 @@ private function convertAmpBindAttributes($html)
+                    ||
+                    in_array($attrName, $this->convertedAmpBindAttributes, true)
+                ) {
+                    $newAttrs .= " [{$attrName}]";
```
Minor point: the replacement includes a single `' '` space. The regex, however, matches any repeated sequence of whitespace. Granted the inter-attribute whitespace at this point would have been normalize by the DOM serialization, but it's just something I wanted to point out. Probably no action required.

schlessera replied:
I'm aware of the discrepancy but opted to keep it for now for the added robustness of accepting more than one whitespace.

westonruter commented on src/Dom/Document.php:
```diff
@@ -1082,7 +1201,93 @@ private function convertAmpBindAttributes($html)
+                    in_array($attrName, $this->convertedAmpBindAttributes, true)
+                ) {
+                    $newAttrs .= " [{$attrName}]";
+                    if (isset($attrMatches['value'])) {
```
Per note above, the value of an AMP-bind attribute should actually always be provided present. A bind attribute by definition has to have a value to control the state which should manage the bound attribute.

schlessera replied:
I'm now falling back to an empty value (`=""`) instead of generating a boolean attribute.

westonruter commented on src/Dom/Document.php:
```diff
@@ -1397,14 +1602,49 @@ private function convertAmpEmojiAttribute($source)
+        if (!preg_match('#^(<html)(\s[^>]+)>$#i', $htmlTag, $matches)) {
+            return $source;
+        }
+
+        // Split into individual attributes.
+        $attributes = array_map(
+            'trim',
+            array_filter(
+                preg_split(
+                    '#(\s+[^"\'\s=]+(?:=(?:"[^"]+"|\'[^\']+\'|[^"\'\s]+))?)#',
+                    $matches[2],
+                    -1,
+                    PREG_SPLIT_DELIM_CAPTURE
+                )
+            )
         );
+
+        foreach ($attributes as $index => $attribute) {
+            $attributeMatches = [];
+            if (
+                preg_match(
+                    '/^(' . Attribute::AMP_EMOJI_ALT . '|' . Attribute::AMP_EMOJI . ')(4(?:ads|email))?$/i',
+                    $attribute,
+                    $attributeMatches
+                )
+            ) {
+                $this->usedAmpEmoji = $attributeMatches[1];
+                $variant            = ! empty($attributeMatches[2]) ? $attributeMatches[2] : '';
+                $attributes[$index] = self::EMOJI_AMP_ATTRIBUTE_PLACEHOLDER . "=\"{$variant}\"";
+
+                $source = preg_replace(
+                    self::AMP_EMOJI_ATTRIBUTE_PATTERN,
+                    '<html ' . implode(' ', $attributes) . '>',
+                    $source
+                );
+                break;
+            }
+        }
+
+        return $source;
```
This could be made more efficient by adapting into a `preg_replace_callback` where the following code is put inside the callback. In that way, there would not be a need to do an additional `preg_replace()` on the `$source`. It could be combined with teh `preg_match()` call above as well. Something like this I believe:

```php
    private function convertAmpEmojiAttribute($source)
    {
        $this->usedAmpEmoji = '';

        return preg_replace_callback(
            self::AMP_EMOJI_ATTRIBUTE_PATTERN,
            static function ( $matches ) {
                // Split into individual attributes.
                $attributes = array_map(
                    'trim',
                    array_filter(
                        preg_split(
                            '#(\s+[^"\'\s=]+(?:=(?:"[^"]+"|\'[^\']+\'|[^"\'\s]+))?)#',
                            $matches[2],
                            -1,
                            PREG_SPLIT_DELIM_CAPTURE
                        )
                    )
                );

                foreach ($attributes as $index => $attribute) {
                    $attributeMatches = [];
                    if (
                        preg_match(
                            '/^(' . Attribute::AMP_EMOJI_ALT . '|' . Attribute::AMP_EMOJI . ')(4(?:ads|email))?$/i',
                            $attribute,
                            $attributeMatches
                        )
                    ) {
                        $this->usedAmpEmoji = $attributeMatches[1];
                        $variant            = ! empty($attributeMatches[2]) ? $attributeMatches[2] : '';
                        $attributes[$index] = self::EMOJI_AMP_ATTRIBUTE_PLACEHOLDER . "=\"{$variant}\"";
                        break;
                    }
                }

                return '<html ' . implode(' ', $attributes) . '>';
            },
            $source,
            1
        );
    }
```

Note also the added limit of 1.

schlessera replied:
I adapted to code to use a single `preg_replace_callback()`. Your code was already pretty close, but it needed a bit more fiddling with the regex pattern(s).

Just to be sure, I profiled the difference across the unit tests we have that cover emojis, and this is the result:

<img width="297" alt="Comparison from Untitled to Untitled - Blackfire - Google Chrome 2021-03-15 at 11 22 19 AM" src="https://user-images.githubusercontent.com/83631/111146532-754a2180-8581-11eb-9203-e36b562bf268.png">

So it's indeed a clear improvement overall.

schlessera commented on src/Dom/Document.php:
```diff
@@ -37,6 +37,42 @@
+     */
+    const OPTION_AMP_BIND_SYNTAX = 'amp_bind_syntax';
+
+    // Different flags for the amp-bind syntax option.
```
While doing so, I decided that it's time to refactor the `Document` class, as it is becoming huge and cumbersome. For now I just extracted some constants into two interfaces `Document\Option` and `Document\Encoding` to not overload this PR. I'll create issues for additional refactoring and deal with them in separate PRs.

schlessera commented on src/Dom/Document.php:
```diff
@@ -87,27 +123,62 @@ final class Document extends DOMDocument
     const AMP_BIND_DATA_ATTR_PREFIX = 'data-amp-bind-';
 
     /**
-     * Pattern for HTML attribute accounting for binding attr name, boolean attribute, single/double-quoted attribute
```
Yes, binding attributes would always need a value. However, I opted to keep the matching intact so as to ensure even bad markup is correctly parsed. I do now enforce adding the value in the conversion below, though.

schlessera commented:
While looking into the few lines Codecov marked as not being tested, I noticed bugs in the encoding code and fixed them in here as well (and added a few tests for them too).

westonruter commented on src/Dom/Document.php:
```diff
@@ -1313,21 +1248,30 @@ public function restoreAmpBindAttributes($html)
+        // In my testing, this was not possible as long as one ISO-8859-x is in the detection order.
         if (empty($this->originalEncoding)) {
-            $this->originalEncoding = self::AMP_ENCODING;
+            $this->originalEncoding = Encoding::AMP;
```
If it's not possible, might as well exclude it from code coverage reporting:

```suggestion
            $this->originalEncoding = Encoding::AMP; // @codeCoverageIgnore
```
